File: pubsubsystem/typescript/subscriber.ts
import { Message } from './message';

export enum SubscriberState {
    Hot = 'hot',
    Cold = 'cold'
}

export class Subscriber {
    name: string;
    state: SubscriberState;

    constructor(name: string, state: SubscriberState) {
        this.name = name;
        this.state = state;
    }

    static newSubscriber(name: string, state: SubscriberState): Subscriber {
        return new Subscriber(name, state);
    }

    onMessage(message: Message): void {
        console.log(`Subscriber ${this.name} received message: ${message.content}`);
    }
}

File: pubsubsystem/subscriber.go
package pubsubsystem

import "fmt"

type SubscriberState string

const (
	Hot  SubscriberState = "hot"
	Cold SubscriberState = "cold"
)

type Subscriber struct {
	Name  string
	State SubscriberState
}

func NewSubscriber(name string, state SubscriberState) *Subscriber {
	return &Subscriber{Name: name, State: state}
}

func (s *Subscriber) OnMessage(message *Message) {
	fmt.Printf("Subscriber %s recieved message: %s\n", s.Name, message.Content)
}
